components/logins/src/login.rs

```diff
@@ -2,6 +2,253 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+//  N.B. if you're making a documentation change here, you might also want to make it in:
+//
+//    * The API docs in ../ios/Logins/LoginRecord.swift
+//    * The API docs in ../android/src/main/java/mozilla/appservices/logins/ServerPassword.kt
+//    * The android-components docs at
+//      https://github.com/mozilla-mobile/android-components/tree/master/components/service/sync-logins
+//
+//  We'll figure out a more scalable approach to maintaining all those docs at some point...
+
+//!
+//! Login Records
+//! =============
+//!
+//! The core datatype managed by this component is a "login record", which contains the following fields:
+//!
+//! - `id`:  A unique string identifier for this record.
+//!
+//!   Consumers may assume that `id` contains only "safe" ASCII characters but should otherwise
+//!   treat this it as an opaque identifier. It should be left blank when adding a new record,
+//!   in which case a new id will be automatically generated.
+//!
+//! - `hostname`:  The origin at which this login can be used, as a string.
+//!
+//!   The login should only be used on sites that match this origin (for whatever definition
+//!   of "matches" makes sense at the application level, e.g. eTLD+1 matching).
+//!   This field is required, must be a valid origin in punycode format, and must not be
+//!   set to the empty string.
+//!
+//!   Examples of valid `hostname` values include:
+//!   - "https://site.com"
+//!   - "http://site.com:1234"
+//!   - "ftp://ftp.site.com"
+//!   - "moz-proxy://127.0.0.1:8888"
+//!   - "chrome://MyLegacyExtension"
+//!   - "file:///"
```

Comment:
I'm not sure if github ate thom's comment or what, but he had a good point here that we need to clarify.

@mnoorenberghe what should the behaviour be for values that have an [opaque origin](https://url.spec.whatwg.org/#origin) per the WhatWG URL Spec?

Concrete example: if one record has `origin="file:///"` and another has `origin="file:///hello/world"`, what happens?

My take is that we should treat them as valid and leave them alone, and that calling code should only consider them a match based on exact string matching.

I had a rummage around in the Desktop `passwordmgr` directory but couldn't find any testcases covering such cases.